### PR TITLE
Use exactly one hse transaction for each Ceph transaction.

### PR DESCRIPTION
**Pull request by pierrelabat**
Abort or commit and free the hse transaction in the top function start_one_transaction()
Use Onode to keep object context during transaction to avoid re-computing the same thing.
Cache collection key in the Collection class to avoid recomputing it.
Implement touch and create operations.
Simplify error code handling (return hse_err_t till to the top function (start_one_transaction()).

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/os/hsestore/HseStore.cc [hunk 1]
```diff
@@ -112,6 +113,7 @@ HseStore::Collection::Collection(HseStore *store, coll_t cid)
   : CollectionImpl(store->cct, cid),
     _store(store)
 {
+  coll_t2key(_store->cct, cid, &_coll_tkey);
```
Comment (tristan957) [inline src/os/hsestore/HseStore.cc:116]: good change

File: src/os/hsestore/HseStore.cc [hunk 2]
```diff
@@ -434,49 +398,27 @@ int HseStore::remove_collection(coll_t cid, CollectionRef *c)
...
-  os.kop_txn = txn;
-
-  rc = hse_kvdb_txn_begin(kvdb, txn);
-  if (rc) {
-    dout(10) << " failed to begin transaction while removing collection("
-      << cid << ')' << dendl;
-    goto err_out;
-  }
-
   coll_t2key(cct, cid, &filt);
```
Comment (tristan957) [inline src/os/hsestore/HseStore.cc:401]: Do you already have access to this from the Collection constructor?

File: src/os/hsestore/HseStore.cc [hunk 3]
```diff
@@ -961,6 +856,38 @@ int HseStore::mkfs()
...
+  Onode& o,
+  const ghobject_t& oid,
+  bool create)
+{
+  hse_oid_t hse_oid;
+  bool found;
+
+  if (!o.o_gotten) {
+    o.o_gotten = true;
+    o.o_oid = std::addressof(oid);
```
Comment (tristan957) [inline src/os/hsestore/HseStore.cc:872]: never knew about this function

File: src/os/hsestore/HseStore.cc [hunk 4]
```diff
@@ -980,32 +907,58 @@ void HseStore::set_fsid(uuid_d u)
...
+  uint32_t fadvise_flags)
+{
+  hse_err_t rc = 0;
+
+  dout(15) << __func__ << " " << c->cid << " " << *o.o_oid
+	   << " " << offset << "~" << length
+	   << dendl;
+  // _assign_nid(txc, o);
+  // int r = _do_write(txc, o, offset, length, bl, fadvise_flags);
+  // txc->write_onode(o);
```
Comment (tristan957) [inline src/os/hsestore/HseStore.cc:925]: Are these comments important?

File: src/os/hsestore/HseStore.cc [hunk 5]
```diff
@@ -1291,43 +1244,44 @@ void HseStore::start_one_transaction(Collection *c, Transaction *t)
...
 	  // by partially applying transactions.
 	  msg = "ENOSPC from key value store, misconfigured cluster";
 
-	if (r == -ENOTEMPTY) {
+	if (rc == ENOTEMPTY) {
 	  msg = "ENOTEMPTY suggests garbage data in osd data dir";
 	}
 
-	dout(0) << " error " << cpp_strerror(r) << " not handled on operation " << op->op
+	dout(0) << " error " << cpp_strerror(rc) << " not handled on operation " << op->op
```
Comment (tristan957) [inline src/os/hsestore/HseStore.cc:1284]: should this include an hse_err_to_errno()

File: src/os/hsestore/HseStore.cc [hunk 6]
```diff
@@ -1717,3 +1737,34 @@ static int key2coll_t(const std::string_view key, coll_t &coll)
   }
   return 0;
 }
+
```
Comment (tristan957) [inline src/os/hsestore/HseStore.cc:1740]: I think it worth encapsulating the structure of the key (describe parts 1, 2, and 3)

File: src/os/hsestore/HseStore.h [hunk 7]
```diff
@@ -128,10 +128,46 @@ class TxnWaitPersist {
...
 
 class HseStore : public ObjectStore {
+  public:
   using hse_oid_t = uint64_t;
+
+  private:
   Syncer *_syncer;
+  //
+  // Next hse_oid to assign to a new Ceph object.
+  std::atomic<uint64_t> _hse_oid;
```
Comment (tristan957) [inline src/os/hsestore/HseStore.h:138]: Rename to _next_hse_oid?

Also if you want to rename the kvdb and kvss to be prefixed with _, let's do that too. Ceph seems to use that pattern in most of the code

**Comment by tristan957** [comment]
Nice PR. Don't forget to rebase on master!

File: src/os/hsestore/HseStore.cc (hunk 2 above)
Comment (pierrelabat) [reply src/os/hsestore/HseStore.cc:401]: Right, will use it.

File: src/os/hsestore/HseStore.cc (hunk 4 above)
Comment (pierrelabat) [reply src/os/hsestore/HseStore.cc:925]: Going to remove them.

File: src/os/hsestore/HseStore.cc (hunk 5 above)
Comment (pierrelabat) [reply src/os/hsestore/HseStore.cc:1284]: Yes.

File: src/os/hsestore/HseStore.cc (hunk 6 above)
Comment (pierrelabat) [reply src/os/hsestore/HseStore.cc:1740]: Don't understand that comment.

File: src/os/hsestore/HseStore.h (hunk 7 above)
Comment (pierrelabat) [reply src/os/hsestore/HseStore.h:138]: Ok, will rename. And change to _...

**Pull request merged**